utils/zerovec/src/varzerovec/components.rs

```diff
@@ -0,0 +1,246 @@
+            })
+            .chain(last)
+            .map(|s| s.and_then(|s| T::VarULE::parse_byte_slice(s).map_err(|e| e.into())))
+    }
+
+    /// Create an iterator over the Ts contained in SliceComponents
+    #[inline]
+    pub fn iter(self) -> impl Iterator<Item = &'a T::VarULE> {
+        let last = iter::from_fn(move || {
+            if self.len() > 0 {
+                let start = usizeify(self.indices[self.len() - 1]);
+                let end = self.things.len();
+                Some(unsafe { self.things.get_unchecked(start..end) })
+            } else {
+                None
+            }
+        })
+        .take(1);
+        self.indices
+            .windows(2)
+            .map(move |win| {
+                let start = usizeify(win[0]);
+                let end = usizeify(win[1]);
+                unsafe { self.things.get_unchecked(start..end) }
+            })
+            .chain(last)
+            .map(|s| unsafe { T::VarULE::from_byte_slice_unchecked(s) })
+    }
+}
+
+impl<'a, T> SliceComponents<'a, T>
+where
+    T: AsVarULE,
+    T::VarULE: Ord,
+{
+    pub fn binary_search(&self, x: &T::VarULE) -> Result<usize, usize> {
+        // This code is an absolute atrocity. This code is not a place of honor. This
+        // code is known to the State of California to cause cancer.
+        //
+        // Unfortunately, the stdlib's `binary_search*` functions can only operate on slices.
+        // We do not have a slice. We have something we can .get() and index on, but that is not
+        // a slice.
+        //
+        // The `binary_search*` functions also do not have a variant where they give you the element's
+        // index, which we could otherwise use to directly index `self`.
+        // We do have `self.indices`, but these are indices into a byte buffer, which cannot in
+        // isolation be used to recoup the logical index of the element they refer to.
+        //
+        // However, `binary_search_by()` provides references to the elements of the slice being iterated.
+        // Since the layout of Rust slices is well-defined, we can do pointer arithmetic on these references
+        // to obtain the index being used by the search.
+        //
+        // It's worth noting that the slice we choose to search is irrelevant, as long as it has the appropriate
+        // length. `self.indices` is defined to have length `self.len()`, so it is convenient to use
+        // here and does not require additional allocations.
+        //
+        // The alternative to doing this is to implement our own binary search. This is significantly less fun.
+
+        let zero_index = self.indices.as_ptr() as *const _ as usize;
+        self.indices.binary_search_by(|probe: &_| {
+            // `self.indices` is a vec of unaligned u32s, so we divide by sizeof(u32)
+            // to get the actual index
+            let index = (probe as *const _ as usize - zero_index) / mem::size_of::<u32>();
```

Comment:
`suggestion(optional)`: can you factor the `mem::size_of` out of the loop? I suspect the compiler is smart enough, but for the off chance it's not...

can you also add to the docs that `var as *const _ as type` is a trick used to cast types arbitrarily and that it's totally safe because we make no mistakes?
Ok, maybe skip the last part.

Reply:
@zbraniecki it's a const fn, so it's _guaranteed_

Reply:
Actually, scratch that, it's not guaranteed, but there is no reason this won't get hoisted